src/include/txn.i

```diff
@@ -508,13 +508,6 @@ __wt_txn_pinned_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *pinned_tsp)
     if (!include_checkpoint_txn)
         return;
 
-    /*
-     * The read of the timestamp pinned by a checkpoint needs to be carefully ordered: if a
-     * checkpoint is starting and we have to use the checkpoint timestamp, we take the minimum of it
-     * with the oldest timestamp, which is what we want.
-     */
-    WT_READ_BARRIER();
```

Comment:
It looks to me like this read barrier is necessary. The comment says that we need to make sure that the read of the checkpoint_timestamp sees the value at least as recent as the read of the pinned_timestamp above. The compiler can re-order across a conditional, so I don't think that helps.

The code would be more obvious if the operations were grouped together. Could the check for `include_checkpoint_txn` be moved above assigning the pinned timestamp? So the two assignments where order matters are grouped together? I'd make it:

```
/*
 * The read of the timestamp pinned by a checkpoint needs to be carefully ordered: if a
 * checkpoint is starting and we have to use the checkpoint timestamp, we take the minimum of it
 * with the oldest timestamp, which is what we want.
 */
*pinned_tsp = pinned_ts = txn_global->pinned_timestamp;
WT_READ_BARRIER();
checkpoint_ts = txn_global->checkpoint_timestamp;
```

That makes the grouping more obvious, so hopefully we won't add code in the middle accidentally again.

Reply:
I think it should also be behind
```
include_checkpoint_txn =
      btree == NULL || (btree->checkpoint_gen != __wt_gen(session, WT_GEN_CHECKPOINT));
```
Otherwise, we may read the checkpoint timestamp from the previous checkpoint.